# Learning Outcomes
**In this tutotial you are going to learn:**

  1: Basic NLP techniques to preprocess text

  2: Differnt techniques to conver text into numerical form (CountVectorize, HashingTF, and TF-IDF)

  3: Apply regression classifier on the text corpus

**Install spark and otehr necessary preliminiries**

In [2]:
# !pip install pyspark

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 317.0/317.0 MB 2.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for pyspark: filename=pyspark-3.5.1-py2.py3-none-any.whl size=317488493 sha256=5025fc944abb55e7ff2fdb3964b564dd40e0916ee69f89f62af69e23a4b3bf57
  Stored in directory: /root/.cache/pip/wheels/80/1d/60/2c256ed38dddce2fdd93be545214a63e02fbd8d74fb0b7f3a6
Successfully built pyspark


In [1]:
from pyspark.sql import SparkSession
spark = SparkSession.builder.master("local[*]").getOrCreate()

In [2]:
spark = SparkSession \
    .builder \
    .appName("Spark_NLP") \
    .config("spark.some.config.option", "some-value") \
    .getOrCreate()

In [4]:
# from google.colab import drive
# drive.mount('/content/drive')

Mounted at /content/drive


#**Steps involve in NLP:**
Text data can be very messy sometimes, and it needs careful attention to bring it to a stage where it can be used in the right way. There are multiple ways in which text data can be cleaned and refined. For example, regular expressions are very powerful when it comes to filtering, cleaning, and standardizing text data. However, regular expressions are not the focus area in this tutorial. Rather, we look at the steps to prepare text data in a form where we can fit a ML model on it. The five major steps involved in handling text data for ML modeling are:

1.   Reading the corpus

2.	Tokenization

3.	Cleaning/stopword removal

4.	Stemming

5.	Converting into numerical form


**Creating Carpous **

A corpus is known as an entire collection of text documents, for example, a collection of emails, messages, or user reviews. This group of individual text items is known as a **corpus**.
We will use CSV file contain collection of text later in this tutorial. First we will do some basic pre-processing using text.

##**1. Tokenization**

The method of dividing the given text sequence/sentence or collection of words of a text document into individual words is known as tokenization. It removes the unnecessary characters such as punctuations. The final units post-tokenization are known as tokens. Let’s say we have the following text:

**Input:** 'He really liked the London City. He is there for two more days'

Tokenization would result in the following tokens.

[He, really, liked, the, London, City, He, is, there, for, two, more, days]

We end up with 13 tokens for the input text:

Let us see how we can do tokenization using PySpark.
The first step is to create a dataframe that has text data.

In [15]:
#create data frame

df2=spark.createDataFrame([(1,'I really liked this movie wathing this movie was an unforgetable experience'),
                         (2,'I would recommend this movie to my friends'),
                         (3,'movie was alright but acting was horrible'),
                         (4,'I am never watching that movie ever again')],
                        ['user_id','review'])


df2.show(4,False)

# When False is passed as the second argument, it means that long strings will not be truncated.

Py4JJavaError: An error occurred while calling o288.showString.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 10.0 failed 1 times, most recent failure: Lost task 0.0 in stage 10.0 (TID 21) (10.210.78.234 executor driver): org.apache.spark.SparkException: Python worker failed to connect back.
	at org.apache.spark.api.python.PythonWorkerFactory.createSimpleWorker(PythonWorkerFactory.scala:203)
	at org.apache.spark.api.python.PythonWorkerFactory.create(PythonWorkerFactory.scala:109)
	at org.apache.spark.SparkEnv.createPythonWorker(SparkEnv.scala:124)
	at org.apache.spark.api.python.BasePythonRunner.compute(PythonRunner.scala:174)
	at org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:67)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:93)
	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:161)
	at org.apache.spark.scheduler.Task.run(Task.scala:141)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$4(Executor.scala:620)
	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally(SparkErrorUtils.scala:64)
	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally$(SparkErrorUtils.scala:61)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:94)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:623)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.net.SocketTimeoutException: Accept timed out
	at java.base/sun.nio.ch.NioSocketImpl.timedAccept(NioSocketImpl.java:701)
	at java.base/sun.nio.ch.NioSocketImpl.accept(NioSocketImpl.java:745)
	at java.base/java.net.ServerSocket.implAccept(ServerSocket.java:698)
	at java.base/java.net.ServerSocket.platformImplAccept(ServerSocket.java:663)
	at java.base/java.net.ServerSocket.implAccept(ServerSocket.java:639)
	at java.base/java.net.ServerSocket.implAccept(ServerSocket.java:585)
	at java.base/java.net.ServerSocket.accept(ServerSocket.java:543)
	at org.apache.spark.api.python.PythonWorkerFactory.createSimpleWorker(PythonWorkerFactory.scala:190)
	... 32 more

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2844)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2780)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2779)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2779)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1242)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1242)
	at scala.Option.foreach(Option.scala:407)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1242)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:3048)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2982)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2971)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:984)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2398)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2419)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2438)
	at org.apache.spark.sql.execution.SparkPlan.executeTake(SparkPlan.scala:530)
	at org.apache.spark.sql.execution.SparkPlan.executeTake(SparkPlan.scala:483)
	at org.apache.spark.sql.execution.CollectLimitExec.executeCollect(limit.scala:61)
	at org.apache.spark.sql.Dataset.collectFromPlan(Dataset.scala:4344)
	at org.apache.spark.sql.Dataset.$anonfun$head$1(Dataset.scala:3326)
	at org.apache.spark.sql.Dataset.$anonfun$withAction$2(Dataset.scala:4334)
	at org.apache.spark.sql.execution.QueryExecution$.withInternalError(QueryExecution.scala:546)
	at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:4332)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$6(SQLExecution.scala:125)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:201)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:108)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:900)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:66)
	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:4332)
	at org.apache.spark.sql.Dataset.head(Dataset.scala:3326)
	at org.apache.spark.sql.Dataset.take(Dataset.scala:3549)
	at org.apache.spark.sql.Dataset.getRows(Dataset.scala:280)
	at org.apache.spark.sql.Dataset.showString(Dataset.scala:315)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:75)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:52)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:374)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
	at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.apache.spark.SparkException: Python worker failed to connect back.
	at org.apache.spark.api.python.PythonWorkerFactory.createSimpleWorker(PythonWorkerFactory.scala:203)
	at org.apache.spark.api.python.PythonWorkerFactory.create(PythonWorkerFactory.scala:109)
	at org.apache.spark.SparkEnv.createPythonWorker(SparkEnv.scala:124)
	at org.apache.spark.api.python.BasePythonRunner.compute(PythonRunner.scala:174)
	at org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:67)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:93)
	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:161)
	at org.apache.spark.scheduler.Task.run(Task.scala:141)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$4(Executor.scala:620)
	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally(SparkErrorUtils.scala:64)
	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally$(SparkErrorUtils.scala:61)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:94)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:623)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	... 1 more
Caused by: java.net.SocketTimeoutException: Accept timed out
	at java.base/sun.nio.ch.NioSocketImpl.timedAccept(NioSocketImpl.java:701)
	at java.base/sun.nio.ch.NioSocketImpl.accept(NioSocketImpl.java:745)
	at java.base/java.net.ServerSocket.implAccept(ServerSocket.java:698)
	at java.base/java.net.ServerSocket.platformImplAccept(ServerSocket.java:663)
	at java.base/java.net.ServerSocket.implAccept(ServerSocket.java:639)
	at java.base/java.net.ServerSocket.implAccept(ServerSocket.java:585)
	at java.base/java.net.ServerSocket.accept(ServerSocket.java:543)
	at org.apache.spark.api.python.PythonWorkerFactory.createSimpleWorker(PythonWorkerFactory.scala:190)
	... 32 more


In this Dataframe, we have four sentences for tokenization. The next step is to import Tokenizer from the **Spark library**. We must then pass the input column and name the output column after tokenization. We use the transform function to apply tokenization to the review column:

In [10]:
from pyspark.ml.feature import Tokenizer
tokenization=Tokenizer(inputCol='review',outputCol='tokens')
tokenized_df2=tokenization.transform(df2)
tokenized_df2.show(4,False)

Py4JJavaError: An error occurred while calling o181.showString.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 5.0 failed 1 times, most recent failure: Lost task 0.0 in stage 5.0 (TID 5) (10.210.78.234 executor driver): org.apache.spark.SparkException: Python worker failed to connect back.
	at org.apache.spark.api.python.PythonWorkerFactory.createSimpleWorker(PythonWorkerFactory.scala:203)
	at org.apache.spark.api.python.PythonWorkerFactory.create(PythonWorkerFactory.scala:109)
	at org.apache.spark.SparkEnv.createPythonWorker(SparkEnv.scala:124)
	at org.apache.spark.api.python.BasePythonRunner.compute(PythonRunner.scala:174)
	at org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:67)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:93)
	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:161)
	at org.apache.spark.scheduler.Task.run(Task.scala:141)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$4(Executor.scala:620)
	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally(SparkErrorUtils.scala:64)
	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally$(SparkErrorUtils.scala:61)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:94)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:623)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.net.SocketTimeoutException: Accept timed out
	at java.base/sun.nio.ch.NioSocketImpl.timedAccept(NioSocketImpl.java:701)
	at java.base/sun.nio.ch.NioSocketImpl.accept(NioSocketImpl.java:745)
	at java.base/java.net.ServerSocket.implAccept(ServerSocket.java:698)
	at java.base/java.net.ServerSocket.platformImplAccept(ServerSocket.java:663)
	at java.base/java.net.ServerSocket.implAccept(ServerSocket.java:639)
	at java.base/java.net.ServerSocket.implAccept(ServerSocket.java:585)
	at java.base/java.net.ServerSocket.accept(ServerSocket.java:543)
	at org.apache.spark.api.python.PythonWorkerFactory.createSimpleWorker(PythonWorkerFactory.scala:190)
	... 32 more

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2844)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2780)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2779)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2779)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1242)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1242)
	at scala.Option.foreach(Option.scala:407)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1242)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:3048)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2982)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2971)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:984)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2398)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2419)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2438)
	at org.apache.spark.sql.execution.SparkPlan.executeTake(SparkPlan.scala:530)
	at org.apache.spark.sql.execution.SparkPlan.executeTake(SparkPlan.scala:483)
	at org.apache.spark.sql.execution.CollectLimitExec.executeCollect(limit.scala:61)
	at org.apache.spark.sql.Dataset.collectFromPlan(Dataset.scala:4344)
	at org.apache.spark.sql.Dataset.$anonfun$head$1(Dataset.scala:3326)
	at org.apache.spark.sql.Dataset.$anonfun$withAction$2(Dataset.scala:4334)
	at org.apache.spark.sql.execution.QueryExecution$.withInternalError(QueryExecution.scala:546)
	at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:4332)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$6(SQLExecution.scala:125)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:201)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:108)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:900)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:66)
	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:4332)
	at org.apache.spark.sql.Dataset.head(Dataset.scala:3326)
	at org.apache.spark.sql.Dataset.take(Dataset.scala:3549)
	at org.apache.spark.sql.Dataset.getRows(Dataset.scala:280)
	at org.apache.spark.sql.Dataset.showString(Dataset.scala:315)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:75)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:52)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:374)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
	at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.apache.spark.SparkException: Python worker failed to connect back.
	at org.apache.spark.api.python.PythonWorkerFactory.createSimpleWorker(PythonWorkerFactory.scala:203)
	at org.apache.spark.api.python.PythonWorkerFactory.create(PythonWorkerFactory.scala:109)
	at org.apache.spark.SparkEnv.createPythonWorker(SparkEnv.scala:124)
	at org.apache.spark.api.python.BasePythonRunner.compute(PythonRunner.scala:174)
	at org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:67)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:93)
	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:161)
	at org.apache.spark.scheduler.Task.run(Task.scala:141)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$4(Executor.scala:620)
	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally(SparkErrorUtils.scala:64)
	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally$(SparkErrorUtils.scala:61)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:94)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:623)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	... 1 more
Caused by: java.net.SocketTimeoutException: Accept timed out
	at java.base/sun.nio.ch.NioSocketImpl.timedAccept(NioSocketImpl.java:701)
	at java.base/sun.nio.ch.NioSocketImpl.accept(NioSocketImpl.java:745)
	at java.base/java.net.ServerSocket.implAccept(ServerSocket.java:698)
	at java.base/java.net.ServerSocket.platformImplAccept(ServerSocket.java:663)
	at java.base/java.net.ServerSocket.implAccept(ServerSocket.java:639)
	at java.base/java.net.ServerSocket.implAccept(ServerSocket.java:585)
	at java.base/java.net.ServerSocket.accept(ServerSocket.java:543)
	at org.apache.spark.api.python.PythonWorkerFactory.createSimpleWorker(PythonWorkerFactory.scala:190)
	... 32 more


We get a new column named tokens that contains the tokens for each sentence.

## **2. Stopword Removal**

As you can observe, the tokens column contains very common words such as “this,” “the,” “to,” “was,” “that,” etc. These words are known as stopwords, and they seem to add very little value to the analysis. If they are to be used in analysis, it increases the computation overheads without adding any value or insight. Hence, it’s preferred to drop these stopwords from the tokens. In PySpark, we use **StopWordsRemover** to remove the stopwords:

In [15]:
from pyspark.ml.feature import StopWordsRemover
stopword_removal=StopWordsRemover(inputCol='tokens',outputCol='refined_tokens')
#We then pass the tokens as the input column and name the output column as refined tokens.
refined_df2=stopword_removal.transform(tokenized_df2)
refined_df2.select(['user_id','tokens','refined_tokens']).show(4,False)

+-------+----------------------------------------------------------------------------------------+----------------------------------------------------------------+
|user_id|tokens                                                                                  |refined_tokens                                                  |
+-------+----------------------------------------------------------------------------------------+----------------------------------------------------------------+
|1      |[i, really, liked, this, movie, wathing, this, movie, was, an, unforgetable, experience]|[really, liked, movie, wathing, movie, unforgetable, experience]|
|2      |[i, would, recommend, this, movie, to, my, friends]                                     |[recommend, movie, friends]                                     |
|3      |[movie, was, alright, but, acting, was, horrible]                                       |[movie, alright, acting, horrible]                              |
|4      |[i, am,

##**3. Stemming/Lemmatization**

Lemmatization is a natural language processing technique that aims to reduce words to their base or dictionary form, known as the lemma. The base form is often referred to as the **root word or the canonical form**. Lemmatization takes into account the context and part of speech (POS) of a word in order to produce the most appropriate base form.

The main goal of lemmatization is to normalize words so that different inflected forms of a word are treated as a single item. For example, the lemma of "running" would be "run," and the lemma of "played" would be "play." By reducing words to their base forms, lemmatization helps to eliminate redundancy and improve the accuracy and efficiency of text analysis tasks.

Lemmatization is different from stemming, another popular text normalization technique. While stemming simply chops off prefixes or suffixes from words to derive the root form, lemmatization considers the vocabulary and morphological analysis of words to determine the appropriate base form.

In natural language processing workflows, lemmatization is often used in tasks such as information retrieval, text classification, topic modeling, and sentiment analysis. It helps to consolidate the representation of words, enabling better understanding and analysis of text data.








In [16]:
#We will use nltk library to define and use lemmatizer.
import nltk
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [17]:
from pyspark.sql.functions import udf
from pyspark.sql.types import ArrayType, StringType
from nltk.stem import WordNetLemmatizer

# Define lemmatization function using nltk library
def lemmatize_text(tokens):
    lemmatizer = WordNetLemmatizer()
    lemmas = [lemmatizer.lemmatize(token) for token in tokens]
    return lemmas

# Register lemmatization function as a UDF
lemmatize_udf = udf(lemmatize_text, ArrayType(StringType()))

# Apply lemmatization using the UDF
refined_df2 = refined_df2.withColumn("lemmatized_words", lemmatize_udf("refined_tokens"))

# Show the resulting DataFrame
refined_df2.show(truncate=False)


+-------+---------------------------------------------------------------------------+----------------------------------------------------------------------------------------+----------------------------------------------------------------+----------------------------------------------------------------+
|user_id|review                                                                     |tokens                                                                                  |refined_tokens                                                  |lemmatized_words                                                |
+-------+---------------------------------------------------------------------------+----------------------------------------------------------------------------------------+----------------------------------------------------------------+----------------------------------------------------------------+
|1      |I really liked this movie wathing this movie was an unforgetable experience|

Note for the second review 'friends' became 'friend'.  When you have bigger text corpous you will observe more changes

##**4.Converting into numerical form/ Feature Engineering**


Now that we have the key tokens created from the text data, we now need a mechanism to convert the tokens into a **numerical form** because we know for a fact that a Machine Learning algorithm works on numerical data. Text data is generally unstructured and varies in its length.There are many techniques and we will use few in this tutorial:


**Bag of Words**

Bag of words allows to convert the text data into numerical vector form by considering the occurrence of the words in text documents, for example:

Doc 1: The best thing in life is to travel

Doc 2: Travel is the best medicine

Doc 3: One should travel more often

**vocabulary**

The list of unique words appearing in all the documents is known as **vocabulary** . In the above example, we have a total of 13 unique words appearing that are part of the vocab as follow.

[The, Best, thing, in, life, is, to, travel, medicine, one, should, more,often]

Any of these three documents can be represented by this vector of fixed size 13 using a Boolean value (1 or 0):

**Doc 1:**
The Best thing in life is to travel medicine one should more often

1   1    1     1  1    1  1  1      0        0    0     0    0

**Doc 2:**

The Best thing in life is to travel medicine one should more often

0   1    0     0   0    0  0   1      1       0    0     0    0


**Doc 3:**

The Best thing in life is to travel medicine one should more often

0   0    0     0  0    0   0   1      0        0    1     1    1

Bag of words does not consider the order of words in the document and the semantic meaning of the word and hence is the most baseline method to represent the text data in numerical form.
There are other ways by which we can convert the textual data into numerical form, which are covered in the following section.
We will use PySpark to go through each one of these methods.

###**CountVectorizer**

In bag of words, we saw the representation of the occurrence of a word by simply **1 or 0** and did not consider the frequency of the word.

The count vectorizer instead takes the **total count of the tokens** appearing in the particular document. We will use the same text documents that we created earlier during tokenization. We first import CountVectorizer:

In [18]:
from pyspark.ml.feature import CountVectorizer
count_vec=CountVectorizer(inputCol='lemmatized_words',outputCol='features')
cv_df2=count_vec.fit(refined_df2).transform(refined_df2)
cv_df2.select(['user_id','lemmatized_words','features']).show(4,False)

+-------+----------------------------------------------------------------+---------------------------------------------+
|user_id|lemmatized_words                                                |features                                     |
+-------+----------------------------------------------------------------+---------------------------------------------+
|1      |[really, liked, movie, wathing, movie, unforgetable, experience]|(14,[0,2,3,5,8,12],[2.0,1.0,1.0,1.0,1.0,1.0])|
|2      |[recommend, movie, friend]                                      |(14,[0,4,11],[1.0,1.0,1.0])                  |
|3      |[movie, alright, acting, horrible]                              |(14,[0,1,9,13],[1.0,1.0,1.0,1.0])            |
|4      |[never, watching, movie, ever]                                  |(14,[0,6,7,10],[1.0,1.0,1.0,1.0])            |
+-------+----------------------------------------------------------------+---------------------------------------------+



CountVectorizer builds a vocabulary of all the unique words in the corpus and assigns each word an index. The dimensionality of the feature vectors is equal to the size of the vocabulary, which can be very large for large corpora.
As we can observe, each row is represented as a dense vector. It shows that the vector length/dimension is 14 and the first sentence contains **6** word/values of the **vocabulary** at the **0,2,3,5, 8 and 12th** indexes.

Note correspond to the 0th index count is 2. It should be value for the word 'movie' in the first review as count of 'movie' in this sentence is 2.
You can use vocabulary function to see number of unique words in the vocabulary



In [19]:
#To validate the vocabulary of the count vectorizer, we can simply use the vocabulary function:
vocabulary=count_vec.fit(refined_df2).vocabulary
vocabulary

['movie',
 'horrible',
 'experience',
 'liked',
 'recommend',
 'really',
 'never',
 'watching',
 'unforgetable',
 'alright',
 'ever',
 'friend',
 'wathing',
 'acting']

Hence, the vocabulary size for the preceding sentences is 11; and if you look at the features carefully, they are like the input feature vector we have been using for Machine Learning in PySpark.

The drawback of using the CountVectorizer method is that it doesn’t consider the **co-occurrences of words in other documents.**

In simple terms, the words appearing often would have larger impact on the feature vector. Hence, another approach to convert text data into numerical form is **TF-IDF** (Term Frequency – Inverse Document Frequency).

###**TF-IDF**

This method tries to normalize the frequency of token occurrence based on other documents. The whole idea is to give more weight to the token if appearing high number of times in the same document but penalize if it is appearing higher number of times in other documents as well. This indicates that the token is common across the corpus and is not as important as its frequency in the current document indicates.

**Term Frequency:** Score based on the frequency of the word in the current document

**Inverse Document Frequency:** Scoring based on the number of documents that contain the current word

Now, we create features based on TF-IDF in PySpark using the same refined df Dataframe:

In [20]:
from pyspark.ml.feature import HashingTF,IDF
hashing_vec2=HashingTF(inputCol='lemmatized_words',outputCol='tf_features')
hashing_df2=hashing_vec2.transform(refined_df2)
hashing_df2.select(['user_id','lemmatized_words','tf_features']).show(4,False)

+-------+----------------------------------------------------------------+---------------------------------------------------------------------------+
|user_id|lemmatized_words                                                |tf_features                                                                |
+-------+----------------------------------------------------------------+---------------------------------------------------------------------------+
|1      |[really, liked, movie, wathing, movie, unforgetable, experience]|(262144,[55875,97005,99172,210223,229264,259272],[1.0,1.0,1.0,2.0,1.0,1.0])|
|2      |[recommend, movie, friend]                                      |(262144,[68228,74520,210223],[1.0,1.0,1.0])                                |
|3      |[movie, alright, acting, horrible]                              |(262144,[95685,171118,210223,236263],[1.0,1.0,1.0,1.0])                    |
|4      |[never, watching, movie, ever]                                  |(262144,[63139,11367

**HashingTF:** It uses a hashing function to map features to indices in a fixed-size feature space. The number of features is determined by the size of the feature space. This means that the dimensionality of the feature vectors is fixed and independent of the actual vocabulary size.

In [22]:
tf_idf_vec2=IDF(inputCol='tf_features',outputCol='tf_idf_features')
tf_idf_df2=tf_idf_vec2.fit(hashing_df2).transform(hashing_df2)
tf_idf_df2.select(['user_id','tf_idf_features']).show(4,False)

+-------+------------------------------------------------------------------------------------------------------------------------------------------------------+
|user_id|tf_idf_features                                                                                                                                       |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------+
|1      |(262144,[55875,97005,99172,210223,229264,259272],[0.9162907318741551,0.9162907318741551,0.9162907318741551,0.0,0.9162907318741551,0.9162907318741551])|
|2      |(262144,[68228,74520,210223],[0.9162907318741551,0.9162907318741551,0.0])                                                                             |
|3      |(262144,[95685,171118,210223,236263],[0.9162907318741551,0.9162907318741551,0.0,0.9162907318741551])                                                  |
|4      |(262144,[63139,113673,203

Now we have required features and we can use machine learning models to do something insightful on the preprocessed text corpous. For example, text classification, clustering and so on.  

#**Text Classification Using Machine Learning**

Now that we understand the steps involved in dealing with text processing and feature vectorization, we can build a text classification model and use it for predictions on text data. The dataset that we are going to use is a sample of the open source movie lens reviews data, and we’re going to predict the sentiment of the given review (positive or negative). Let’s start with reading the text data first and creating a Spark Dataframe:

In [23]:
 # Read CSV file into a DataFrame
 #download the CSV file came along with this tutorial
 #save in ggogle drive and change path according tou your dierectory structure
 text_df= spark.read.csv('Movie_reviews.csv',inferSchema=True, header =True)

**Exploratory Data Analysis (EDA)**

Explore data to understand it by looking it number of rows, columns, data types, etc.  

In [24]:
#check number of columns and their data type
text_df.printSchema()

root
 |-- Review: string (nullable = true)
 |-- Sentiment: string (nullable = true)



As we can see, the Sentiment column is StringType, and we will need to convert it into an integer or float type going forward:

In [25]:

# Show the first few rows of the DataFrame
text_df.show(10, False)

+------------------------------------------------------------------------+---------+
|Review                                                                  |Sentiment|
+------------------------------------------------------------------------+---------+
|The Da Vinci Code book is just awesome.                                 |1        |
|this was the first clive cussler i've ever read, but even books like Rel|1        |
|i liked the Da Vinci Code a lot.                                        |1        |
|i liked the Da Vinci Code a lot.                                        |1        |
|I liked the Da Vinci Code but it ultimatly didn't seem to hold it's own.|1        |
|that's not even an exaggeration ) and at midnight we went to Wal-Mart to|1        |
|I loved the Da Vinci Code, but now I want something better and different|1        |
|i thought da vinci code was great, same with kite runner.               |1        |
|The Da Vinci Code is actually a good movie...                   

In [27]:
print((text_df.count(), len(text_df.columns)))

(7087, 2)


We have close to 7k records, out of which some might not be labeled properly. Hence, we filter only those records that are labeled correctly:

In [30]:
text_df=text_df.filter(((text_df.Sentiment =='1') | (text_df.Sentiment =='0')))
text_df.count()

6990

Some of the records got filtered out, and we are now left with 6990 records for the analysis. The next step is to validate the number of reviews for each class:

In [31]:
text_df.select('Sentiment').distinct().count()

2

In [32]:
#We have no unlabeled tweets in the dataset. The next step is to validate the number of reviews for each class:
text_df.groupBy('Sentiment').count().show()

+---------+-----+
|Sentiment|count|
+---------+-----+
|        0| 3081|
|        1| 3909|
+---------+-----+



We are dealing with a **balanced dataset**  here as both classes have an almost similar number of reviews.

As a next step, we create a new integer-type **Label column** and **drop the original Sentiment column,** which was string type:

In [33]:
text_df=text_df.withColumn("Label",text_df.Sentiment.cast('float')).drop('Sentiment')
from pyspark.sql.functions import rand
text_df.orderBy(rand()).show(10,False)

+-----------------------------------------------------------------------+-----+
|Review                                                                 |Label|
+-----------------------------------------------------------------------+-----+
|Brokeback Mountain is a really depressing movie...                     |0.0  |
|Brokeback Mountain was an AWESOME movie.                               |1.0  |
|Da Vinci Code sucks.                                                   |0.0  |
|Da Vinci Code sucks be...                                              |0.0  |
|Brokeback Mountain was so awesome.                                     |1.0  |
|Is it just me, or does Harry Potter suck?...                           |0.0  |
|we're gonna like watch Mission Impossible or Hoot.(                    |1.0  |
|we're gonna like watch Mission Impossible or Hoot.(                    |1.0  |
|da vinci code was an awesome movie...                                  |1.0  |
|i love being a sentry for mission impos

We include an additional column that captures the length of the review:

In [34]:
from pyspark.sql.functions import length
from pyspark.sql.functions import rand
text_df=text_df.withColumn('length',length(text_df['Review']))
text_df.orderBy(rand()).show(10,False)

+-------------------------------------------------------------------+-----+------+
|Review                                                             |Label|length|
+-------------------------------------------------------------------+-----+------+
|The Da Vinci Code is awesome..                                     |1.0  |30    |
|Da Vinci Code sucked..                                             |0.0  |22    |
|* brokeback mountain is an awesome movie..                         |1.0  |42    |
|da vinci code sucks...                                             |0.0  |22    |
|I hate Harry Potter.                                               |0.0  |20    |
|by the way, the Da Vinci Code sucked, just letting you know...     |0.0  |62    |
|dudeee i LOVED brokeback mountain!!!!                              |1.0  |37    |
|Ok brokeback mountain is such a horrible movie.                    |0.0  |47    |
|I love Harry Potter.                                               |1.0  |20    |
|the

In [35]:
text_df.groupBy('Label').agg({'Length':'mean'}).show()

+-----+-----------------+
|Label|      avg(Length)|
+-----+-----------------+
|  1.0|47.61882834484523|
|  0.0|50.95845504706264|
+-----+-----------------+



There is no major difference between the average length of the positive and negative review.

**Tokenization and Stopwords Removing**

The next step is to start the **tokenization process and remove stopwords:**

In [36]:
tokenization=Tokenizer(inputCol='Review',outputCol='tokens')
tokenized_df=tokenization.transform(text_df)
tokenized_df.show(10, False)


+------------------------------------------------------------------------+-----+------+----------------------------------------------------------------------------------------+
|Review                                                                  |Label|length|tokens                                                                                  |
+------------------------------------------------------------------------+-----+------+----------------------------------------------------------------------------------------+
|The Da Vinci Code book is just awesome.                                 |1.0  |39    |[the, da, vinci, code, book, is, just, awesome.]                                        |
|this was the first clive cussler i've ever read, but even books like Rel|1.0  |72    |[this, was, the, first, clive, cussler, i've, ever, read,, but, even, books, like, rel] |
|i liked the Da Vinci Code a lot.                                        |1.0  |32    |[i, liked, the, da, vinci, c

In [38]:
 #OR use RegexTokenizer to only keep the words and remove the special characters such as ')',etc
 # Remove hashtags and special characters from the review column
 from pyspark.ml.feature import RegexTokenizer
 tokenization=RegexTokenizer(inputCol= 'Review' , outputCol= 'tokens', pattern= '\\W')
 tokenized_df=tokenization.transform(text_df)
tokenized_df.show(10, False)

+------------------------------------------------------------------------+-----+------+-----------------------------------------------------------------------------------------+
|Review                                                                  |Label|length|tokens                                                                                   |
+------------------------------------------------------------------------+-----+------+-----------------------------------------------------------------------------------------+
|The Da Vinci Code book is just awesome.                                 |1.0  |39    |[the, da, vinci, code, book, is, just, awesome]                                          |
|this was the first clive cussler i've ever read, but even books like Rel|1.0  |72    |[this, was, the, first, clive, cussler, i, ve, ever, read, but, even, books, like, rel]  |
|i liked the Da Vinci Code a lot.                                        |1.0  |32    |[i, liked, the, da, vin

In [39]:
#Stop word removal
from pyspark.ml.feature import StopWordsRemover
stopword_removal=StopWordsRemover(inputCol='tokens',outputCol='refined_tokens')
refined_text_df=stopword_removal.transform(tokenized_df)
refined_text_df.show(10, False)
#note the difeerence between tokens and refined_tokens to understand how stop word removal works

+------------------------------------------------------------------------+-----+------+-----------------------------------------------------------------------------------------+---------------------------------------------------------------+
|Review                                                                  |Label|length|tokens                                                                                   |refined_tokens                                                 |
+------------------------------------------------------------------------+-----+------+-----------------------------------------------------------------------------------------+---------------------------------------------------------------+
|The Da Vinci Code book is just awesome.                                 |1.0  |39    |[the, da, vinci, code, book, is, just, awesome]                                          |[da, vinci, code, book, awesome]                               |
|this was the first clive cussle

Since we are now dealing with tokens only instead of an entire review, it would make more sense to capture a number of tokens in each review rather than using the length of the review. We create another column (token count) that gives the number of tokens in each row.

In [40]:
from pyspark.sql.functions import udf
from pyspark.sql.types import IntegerType
from pyspark.sql.functions import *

In [41]:
len_udf = udf(lambda s: len(s), IntegerType())
#This is a user defined function (udf), the lambda function takes a single argument s,
#which is a string. It calculates the length of the string s using the built-in len() function
#return an integer value for length of the string.
refined_text_df = refined_text_df.withColumn("token_count", len_udf(col('refined_tokens')))

In [42]:
refined_text_df.orderBy(rand()).show(10)

+--------------------+-----+------+--------------------+--------------------+-----------+
|              Review|Label|length|              tokens|      refined_tokens|token_count|
+--------------------+-----+------+--------------------+--------------------+-----------+
|looks amazingly f...|  1.0|    72|[looks, amazingly...|[looks, amazingly...|          8|
|Combining the opi...|  0.0|    71|[combining, the, ...|[combining, opini...|          9|
|Also, Da Vinci Co...|  0.0|    35|[also, da, vinci,...|[also, da, vinci,...|          5|
|After some frenzi...|  1.0|    72|[after, some, fre...|[frenzied, phone,...|          7|
|The Da Vinci Code...|  1.0|    41|[the, da, vinci, ...|[da, vinci, code,...|          5|
|I hate Harry Pott...|  0.0|    71|[i, hate, harry, ...|[hate, harry, pot...|          7|
|Then snuck into B...|  0.0|    72|[then, snuck, int...|[snuck, brokeback...|          5|
|Harry Potter drag...|  0.0|    69|[harry, potter, d...|[harry, potter, d...|          8|
|Then snuc

**Converting into numerical form/ Feature Engineering**

Now that we have the refined tokens after stopword removal, we can use any of the preceding approaches to convert text into numerical features. In this case, we use CountVectorizer for feature vectorization for the Machine Learning model:

In [43]:
from pyspark.ml.feature import CountVectorizer
count_vec=CountVectorizer(inputCol='refined_tokens',outputCol='features')
cv_text_df=count_vec.fit(refined_text_df).transform(refined_text_df)
cv_text_df.select(['refined_tokens','features','Label']).show(10, False)

+---------------------------------------------------------------+----------------------------------------------------------------------------------------+-----+
|refined_tokens                                                 |features                                                                                |Label|
+---------------------------------------------------------------+----------------------------------------------------------------------------------------+-----+
|[da, vinci, code, book, awesome]                               |(1644,[0,1,2,8,167],[1.0,1.0,1.0,1.0,1.0])                                              |1.0  |
|[first, clive, cussler, ve, ever, read, even, books, like, rel]|(1644,[11,41,42,178,180,206,214,806,990,1434],[1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0])|1.0  |
|[liked, da, vinci, code, lot]                                  |(1644,[0,1,2,43,182],[1.0,1.0,1.0,1.0,1.0])                                             |1.0  |
|[liked, da, vinci, code, lot]    

(1644,[0,1,2,8,167],[1.0,1.0,1.0,1.0,1.0])

Vocabulary size/dimension is 1644.
Each row is represented as a vector. It shows that the vector length is 1644 and the first review contains five values at the 0,1,2,8,and 167th indexes.All other values shoud be zero

In [44]:
#explore vocabulary
vocabulary=count_vec.fit(refined_text_df).vocabulary
vocabulary

['da',
 'vinci',
 'code',
 'harry',
 'brokeback',
 'mountain',
 'potter',
 'love',
 'awesome',
 'mission',
 'impossible',
 'like',
 'movie',
 'hate',
 'sucked',
 'sucks',
 'much',
 'really',
 'movies',
 'know',
 'suck',
 '3',
 'want',
 'loved',
 'think',
 'one',
 'stupid',
 'depressing',
 '2',
 'reading',
 'horrible',
 'fucking',
 'terrible',
 'oh',
 'right',
 'left',
 'ok',
 'felicia',
 'beautiful',
 'went',
 'saw',
 'read',
 'first',
 'liked',
 'tom',
 'absolutely',
 'way',
 'heard',
 'big',
 'time',
 'film',
 'said',
 'going',
 'boring',
 'series',
 'watch',
 'great',
 'people',
 'yeah',
 'man',
 're',
 'got',
 'better',
 'watching',
 'story',
 'things',
 'b',
 'last',
 'friday',
 'gay',
 'person',
 'wait',
 'rocks',
 'says',
 'cool',
 'anyone',
 'excellent',
 'always',
 'knows',
 'anyway',
 'mom',
 'friends',
 'review',
 'worth',
 'daniel',
 'opinion',
 'dad',
 'either',
 'care',
 'thats',
 'never',
 'p',
 'guy',
 'stand',
 'luv',
 'gonna',
 'table',
 'awards',
 'hates',
 'head',
 

In [46]:
# Now we have a feture vector for each row and ready for the next step to make prediction.
#we only use feature vector and Lable to train any ML model, hence select these two columns from the data frame.
model_text_df=cv_text_df.select(['features','Label'])
model_text_df.show()

+--------------------+-----+
|            features|Label|
+--------------------+-----+
|(1644,[0,1,2,8,16...|  1.0|
|(1644,[11,41,42,1...|  1.0|
|(1644,[0,1,2,43,1...|  1.0|
|(1644,[0,1,2,43,1...|  1.0|
|(1644,[0,1,2,43,2...|  1.0|
|(1644,[39,178,639...|  1.0|
|(1644,[0,1,2,22,2...|  1.0|
|(1644,[0,1,2,56,1...|  1.0|
|(1644,[0,1,2,12,1...|  1.0|
|(1644,[0,1,2,167,...|  1.0|
|(1644,[0,1,2,18,2...|  1.0|
|(1644,[0,1,2,167,...|  1.0|
|(1644,[0,1,2,211,...|  1.0|
|(1644,[0,1,2,17,1...|  1.0|
|(1644,[0,1,2,7],[...|  1.0|
|(1644,[0,1,2,23],...|  1.0|
|(1644,[0,1,2,38,2...|  1.0|
|(1644,[0,1,2,8,16...|  1.0|
|(1644,[0,1,2,200,...|  1.0|
|(1644,[0,1,2,188,...|  1.0|
+--------------------+-----+
only showing top 20 rows



Once we have the feature vector for each row, we can make use of VectorAssembler to create input features for the Machine Learning model:


In [47]:
from pyspark.ml.feature import VectorAssembler
df_assembler = VectorAssembler(inputCols=['features'],outputCol='features_vec')
model_text_df = df_assembler.transform(model_text_df)
model_text_df.printSchema()

root
 |-- features: vector (nullable = true)
 |-- Label: float (nullable = true)
 |-- features_vec: vector (nullable = true)



In [48]:
model_text_df.show()

+--------------------+-----+--------------------+
|            features|Label|        features_vec|
+--------------------+-----+--------------------+
|(1644,[0,1,2,8,16...|  1.0|(1644,[0,1,2,8,16...|
|(1644,[11,41,42,1...|  1.0|(1644,[11,41,42,1...|
|(1644,[0,1,2,43,1...|  1.0|(1644,[0,1,2,43,1...|
|(1644,[0,1,2,43,1...|  1.0|(1644,[0,1,2,43,1...|
|(1644,[0,1,2,43,2...|  1.0|(1644,[0,1,2,43,2...|
|(1644,[39,178,639...|  1.0|(1644,[39,178,639...|
|(1644,[0,1,2,22,2...|  1.0|(1644,[0,1,2,22,2...|
|(1644,[0,1,2,56,1...|  1.0|(1644,[0,1,2,56,1...|
|(1644,[0,1,2,12,1...|  1.0|(1644,[0,1,2,12,1...|
|(1644,[0,1,2,167,...|  1.0|(1644,[0,1,2,167,...|
|(1644,[0,1,2,18,2...|  1.0|(1644,[0,1,2,18,2...|
|(1644,[0,1,2,167,...|  1.0|(1644,[0,1,2,167,...|
|(1644,[0,1,2,211,...|  1.0|(1644,[0,1,2,211,...|
|(1644,[0,1,2,17,1...|  1.0|(1644,[0,1,2,17,1...|
|(1644,[0,1,2,7],[...|  1.0|(1644,[0,1,2,7],[...|
|(1644,[0,1,2,23],...|  1.0|(1644,[0,1,2,23],...|
|(1644,[0,1,2,38,2...|  1.0|(1644,[0,1,2,38,2...|


**We can use any of the classification model on this data, but we proceed with training a logistic regression model:**

In [49]:
from pyspark.ml.classification import LogisticRegression
training_df,test_df=model_text_df.randomSplit([0.75,0.25])

In [50]:
#To verify the presence of enough records for both classes in train and test sets, we can apply the groupBy function on the Label column:
training_df.groupBy('Label').count().show()

+-----+-----+
|Label|count|
+-----+-----+
|  1.0| 2948|
|  0.0| 2300|
+-----+-----+



In [51]:
test_df.groupBy('Label').count().show()

+-----+-----+
|Label|count|
+-----+-----+
|  1.0|  961|
|  0.0|  781|
+-----+-----+



In [52]:
log_reg=LogisticRegression(featuresCol='features_vec',labelCol='Label').fit(training_df)

After training the model, we evaluate the performance of the model on the **test dataset.**

In [53]:
results=log_reg.evaluate(test_df).predictions
results.show()

+--------------------+-----+--------------------+--------------------+--------------------+----------+
|            features|Label|        features_vec|       rawPrediction|         probability|prediction|
+--------------------+-----+--------------------+--------------------+--------------------+----------+
|(1644,[0,1,2,7],[...|  1.0|(1644,[0,1,2,7],[...|[-23.163802820220...|[8.71140600991758...|       1.0|
|(1644,[0,1,2,7],[...|  1.0|(1644,[0,1,2,7],[...|[-23.163802820220...|[8.71140600991758...|       1.0|
|(1644,[0,1,2,7],[...|  1.0|(1644,[0,1,2,7],[...|[-23.163802820220...|[8.71140600991758...|       1.0|
|(1644,[0,1,2,7],[...|  1.0|(1644,[0,1,2,7],[...|[-23.163802820220...|[8.71140600991758...|       1.0|
|(1644,[0,1,2,7],[...|  1.0|(1644,[0,1,2,7],[...|[-23.163802820220...|[8.71140600991758...|       1.0|
|(1644,[0,1,2,7],[...|  1.0|(1644,[0,1,2,7],[...|[-23.163802820220...|[8.71140600991758...|       1.0|
|(1644,[0,1,2,7],[...|  1.0|(1644,[0,1,2,7],[...|[-23.163802820220...|[8.

In [54]:
#Evaluate the model performance using confusion matrix
true_postives = results[(results.Label == 1) & (results.prediction == 1)].count()
true_negatives = results[(results.Label == 0) & (results.prediction == 0)].count()
false_positives = results[(results.Label == 0) & (results.prediction == 1)].count()
false_negatives = results[(results.Label == 1) & (results.prediction == 0)].count()

In [60]:
print ('TP=', true_postives)
print('TN =', true_negatives)
print('FP =', false_positives)
print('FN =', false_negatives)

TP= 957
TN = 762
FP = 19
FN = 4


In [61]:
recall = float(true_postives)/(true_postives + false_negatives)
print(recall)

0.9958376690946931


In [62]:
precision = float(true_postives) / (true_postives + false_positives)
print(precision)

0.9805327868852459


In [63]:
accuracy=float((true_postives+true_negatives) /(results.count()))
print(accuracy)

0.986796785304248


**What is Next?**
We used CountVectorizer to create features for classification model. You can use TF-ID and Observe the difference in results using CountVectorizer and TF-IDF model.

Use different classification models which you have learned earlier and identify the bestt model using different feature engineering techniques such as TF-IDF, CountVectorizer.

**Next Lesson**

In the next session we will continue with other text represenation techniques, such as word2vec

In [72]:
!jupyter nbconvert --to html "/content/drive/MyDrive/ML&Bigdata/CN7030-Feature_Extraction using_TF-IDF.ipynb"

[NbConvertApp] Converting notebook /content/drive/MyDrive/ML&Bigdata/CN7030-Feature_Extraction using_TF-IDF.ipynb to html
[NbConvertApp] Writing 722895 bytes to /content/drive/MyDrive/ML&Bigdata/CN7030-Feature_Extraction using_TF-IDF.html
